lib/cib/cib_utils.c

```diff
@@ -77,6 +78,128 @@ cib_diff_version_details(xmlNode * diff, int *admin_epoch, int *epoch, int *upda
     return TRUE;
 }
 
+#define XPATH_DIFF_V1 "//" F_CIB_UPDATE_RESULT "//" XML_TAG_DIFF_ADDED
+
+/*!
+ * \internal
+ * \brief Check whether a given CIB element was modified in a CIB update (v1)
+ *
+ * Search a CIB diff notification message's XML v1 patchset for the given
+ * element.
+ *
+ * \param[in] patchset  XML patchset from CIB update
+ * \param[in] element   XML tag of CIB element to check (\c NULL is equivalent
+ *                      to \c XML_TAG_CIB)
+ *
+ * \return \c true if \p element was modified in the CIB update, or \c false
+ *         otherwise
+ */
+static bool
+element_in_patchset_v1(const xmlNode *patchset, const char *element)
+{
+    char *xpath = crm_strdup_printf(XPATH_DIFF_V1 "//%s",
+                                    pcmk__s(element, XML_TAG_CIB));
+    xmlXPathObject *xpath_obj = xpath_search(patchset, xpath);
+
+    free(xpath);
+
+    if (xpath_obj == NULL) {
+        return false;
+    }
+    freeXpathObject(xpath_obj);
+    return true;
+}
+
+/*!
+ * \internal
+ * \brief Check whether a given CIB element was modified in a CIB update (v2)
+ *
+ * Search a CIB diff notification message's XML v2 patchset for the given
+ * element.
+ *
+ * \param[in] patchset  XML patchset from CIB update
+ * \param[in] element   XML tag of CIB element to check (\c NULL is equivalent
+ *                      to \c XML_TAG_CIB). Supported values include any CIB
+ *                      element supported by \c pcmk__cib_abs_xpath_for().
+ *
+ * \return \c true if \p element was modified in the CIB update, or \c false
+ *         otherwise
+ */
+static bool
+element_in_patchset_v2(const xmlNode *patchset, const char *element)
+{
+    const char *element_xpath = pcmk__cib_abs_xpath_for(element);
+    const char *parent_xpath = pcmk_cib_parent_name_for(element);
+
+    CRM_CHECK(element_xpath != NULL, return false); // Unsupported element
+
+    for (const xmlNode *change = first_named_child(patchset, XML_DIFF_CHANGE);
+         change != NULL; change = crm_next_same_xml(change)) {
+
+        const char *op = crm_element_value(change, F_CIB_OPERATION);
+        const char *diff_xpath = crm_element_value(change, XML_DIFF_PATH);
+
+        if (pcmk__str_eq(op, "create", pcmk__str_none)) {
+            // Newly added element
+            if (pcmk__str_eq(diff_xpath, parent_xpath, pcmk__str_none)
+                && pcmk__xe_is(pcmk__xml_first_child(change), element)) {
+
+                return true;
+            }
+
+        } else if (pcmk__starts_with(diff_xpath, element_xpath)) {
```

Comment:
On second thought, this probably should be an `if` rather than an `else if`.

Suppose we're concerned about `/elem1`. Suppose we have a create operation where `/elem1/elem2/elem3` is created (as a child of `/elem1/elem2`. I don't think we'd hit any of these conditions. But `/elem1` did change.